File: Days 51 - 60/Day 54/SudokuSolver.cs
using System;

namespace DailyCodingProblem
{
	internal class Day54
	{
		private const int BoxSize = 9;
		private const int SubBoxSize = 3;
		private const int EmptyCell = 0;

		private static int Main(string[] args)
		{
			int[,] sudoku = new int[BoxSize, BoxSize]
			{
				{ 0, 4, 3, 2, 6, 0, 0, 0, 0 },
				{ 0, 0, 6, 0, 0, 0, 0, 0, 3 },
				{ 0, 8, 9, 1, 0, 5, 0, 0, 0 },
				{ 0, 3, 0, 6, 0, 0, 0, 0, 4 },
				{ 4, 7, 0, 0, 1, 0, 0, 3, 8 },
				{ 8, 0, 0, 0, 0, 4, 0, 7, 0 },
				{ 0, 0, 0, 7, 0, 6, 1, 8, 0 },
				{ 1, 0, 0, 0, 0, 0, 9, 0, 0 },
				{ 0, 0, 0, 0, 9, 1, 3, 2, 0 }
			};

			SolveSudoku(sudoku);
			PrintSudoku(sudoku);

			Console.ReadLine();

			return 0;
		}

		private static bool SolveSudoku(int[,] grid)
		{
			int row = 0;
			int column = 0;

			if (!FindEmptyCell(grid, ref row, ref column))
			{
				return true;
			}

			for (int number = 1; number <= BoxSize; number++)
			{
				if (IsValidNumber(grid, row, column, number))
				{
					grid[row, column] = number;

					if (SolveSudoku(grid))
					{
						return true;
					}
					else
					{
						grid[row, column] = EmptyCell;
					}
				}
			}

			return false;
		}

		private static bool IsValidNumber(int[,] grid, int row, int column, int number)
		{
			for (int i = 0; i < BoxSize; i++)
			{
				if (grid[row, i] == number || grid[i, column] == number)
				{
					return false;
				}
			}

			for (int r = 0; r < SubBoxSize; r++)
			{
				for (int c = 0; c < SubBoxSize; c++)
				{
					if (grid[r + row - (row % SubBoxSize), c + column - (column % SubBoxSize)] == number)
					{
						return false;
					}
				}
			}

			return grid[row, column] == EmptyCell;
		}

		private static bool FindEmptyCell(int[,] grid, ref int row, ref int column)
		{
			for (row = 0; row < BoxSize; row++)
			{
				for (column = 0; column < BoxSize; column++)
				{
					if (grid[row, column] == EmptyCell)
					{
						return true;
					}
				}
			}

			return false;
		}

		private static void PrintSudoku(int[,] grid)
		{
			for (int i = 0; i < grid.GetLength(0); i++)
			{
				for (int j = 0; j < grid.GetLength(1); j++)
				{
					Console.Write($"{grid[i, j]} ");
				}

				Console.WriteLine();
			}
		}
	}
}

File: Days 051 - 060/Day 54/SudokuSolver.cpp
#include <array>
#include <iostream>

static constexpr unsigned int BOX_SIZE = 9;
static constexpr unsigned int SUB_BOX_SIZE = 3;
static constexpr unsigned int EMPTY_CELL = 0;

bool FindEmptyCell(const std::array<std::array<unsigned int, BOX_SIZE>, BOX_SIZE>& grid, unsigned int& row, unsigned int& column) noexcept
{
	for (row = 0; row < BOX_SIZE; row++)
	{
		for (column = 0; column < BOX_SIZE; column++)
		{
			if (grid[row][column] == EMPTY_CELL)
			{
				return true;
			}
		}
	}

	return false;
}

bool IsValidNumber(const std::array<std::array<unsigned int, BOX_SIZE>, BOX_SIZE>& grid, unsigned int row, unsigned int column, unsigned int number) noexcept
{
	for (unsigned int i = 0; i < BOX_SIZE; c++)
	{
		if (grid[row][i] == number || grid[i][column] == number)
		{
			return false;
		}
	}

	for (unsigned int r = 0; r < SUB_BOX_SIZE; r++)
	{
		for (unsigned int c = 0; c < SUB_BOX_SIZE; c++)
		{
			if (grid[r + row - (row % SUB_BOX_SIZE)][c + column - (column % SUB_BOX_SIZE)] == number)
			{
				return false;
			}
		}
	}

	return grid[row][column] == EMPTY_CELL;
}

bool SolveSudoku(std::array<std::array<unsigned int, BOX_SIZE>, BOX_SIZE>& grid) noexcept
{
	unsigned int row;
	unsigned int column;

	if (!FindEmptyCell(grid, row, column))
	{
		return true;
	}

	for (unsigned int number = 1; number <= BOX_SIZE; number++)
	{
		if (IsValidNumber(grid, row, column, number))
		{
			grid[row][column] = number;

			if (SolveSudoku(grid))
			{
				return true;
			}
			else
			{
				grid[row][column] = EMPTY_CELL;
			}
		}
	}

	return false;
}

void PrintSudoku(std::array<std::array<unsigned int, BOX_SIZE>, BOX_SIZE>& grid)
{
	for (const auto& row : grid)
	{
		for (const int cell : row)
		{
			std::cout << cell << " ";
		}

		std::cout << "\n";
	}
}

int main(int argc, const char* argv[])
{
	std::array<std::array<unsigned int, BOX_SIZE>, BOX_SIZE> sudoku
	{
		std::array<unsigned int, BOX_SIZE> { 0, 4, 3, 2, 6, 0, 0, 0, 0 },
		std::array<unsigned int, BOX_SIZE> { 0, 0, 6, 0, 0, 0, 0, 0, 3 },
		std::array<unsigned int, BOX_SIZE> { 0, 8, 9, 1, 0, 5, 0, 0, 0 },
		std::array<unsigned int, BOX_SIZE> { 0, 3, 0, 6, 0, 0, 0, 0, 4 },
		std::array<unsigned int, BOX_SIZE> { 4, 7, 0, 0, 1, 0, 0, 3, 8 },
		std::array<unsigned int, BOX_SIZE> { 8, 0, 0, 0, 0, 4, 0, 7, 0 },
		std::array<unsigned int, BOX_SIZE> { 0, 0, 0, 7, 0, 6, 1, 8, 0 },
		std::array<unsigned int, BOX_SIZE> { 1, 0, 0, 0, 0, 0, 9, 0, 0 },
		std::array<unsigned int, BOX_SIZE> { 0, 0, 0, 0, 9, 1, 3, 2, 0 }
	};
	
	SolveSudoku(sudoku);
	PrintSudoku(sudoku);

	std::cin.get();

	return 0;
}